### PR TITLE
FLINK-5803-Add [partitioned] processing time OVER RANGE BETWEEN UNBOUNDED PRECEDING aggregation to SQL

### PR DESCRIPTION
**Pull request by sunjincheng121**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/datastream/DataStreamOverAggregateRule.scala [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    val aggregateCalls = overWindow.getAggregateCalls(logicWindow)
+
+    val namedAggregates = for (i <- 0 until aggregateCalls.size())
+      yield new CalcitePair[AggregateCall, String](aggregateCalls.get(i), "w0$o" + i)
+
+    // both ROWS and RANGE clause with UNBOUNDED PRECEDING and CURRENT ROW condition.
+    if (overWindow.lowerBound.isUnbounded &&  overWindow.upperBound.isCurrentRow) {
+      val name = Option(ExpressionParser.parseExpression(logicWindow.getRelTypeName))
+      val preceding = ExpressionParser.parseExpression("UNBOUNDED")
+      val following = ExpressionParser.parseExpression("CURRENT")
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/datastream/DataStreamOverAggregateRule.scala]: 我不知道这里是否有更好的选择，谁review代码帮我想一想.谢谢

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+      case _ =>
+        throw TableException("other type over Aggregation is currently not supported...")
+    }
+    result
+  }
+}
+
+object DataStreamOverAggregate{
+
+  private def isProcessingUnboundedAndCurrentRow(
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala]: 我不知道这里是否有更好的选择，谁review代码帮我想一想.谢谢

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        getRowType,
+        namedAggregates,
+        namedProperties))
+  }
+
+  override def translateToPlan(tableEnv: StreamTableEnvironment): DataStream[Row] = {
+
+    val partitionKeys = partition.toArray
+
+    val projectionsExceptAggregates = (for(i <- 0 until inputType.getFieldCount) yield i).toArray
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala]: 好像这是bug

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala (hunk 2 above)
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala]: 是否可以定义一个表示边界的Enum，比如
enum RowBoundary {
  UNBOUNDED,
CURRENT,
RELATIVE
}

然后提供一个方法解析边界，返回上界和下界（low，high），外面使用match来处理你想处理的情况
比如这种情况下是：
case (UNBOUNDED, CURRENT) => ....

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/datastream/DataStreamOverAggregateRule.scala (hunk 1 above)
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/datastream/DataStreamOverAggregateRule.scala]: 这parse出来的是个什么东西

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala (hunk 2 above)
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala]: 或者你也可以让ExpressionParser将这几个关键字parse成一个统一的case class来替代enum的作用，然后这里也可以使用pattern match来做，和enum道理一样。

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/datastream/DataStreamOverAggregateRule.scala (hunk 1 above)
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/datastream/DataStreamOverAggregateRule.scala]: 我将判断后移了，这样简洁多了.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala (hunk 2 above)
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala]: 恩，我仔细想了一下，我将判断后移了，这样简洁多了.

**Pull request closed without merging**